docs/extensions/storage-transformers/sharding/v1.0.rst

```diff
@@ -0,0 +1,293 @@
+  ``get_partial_values``. This operation should be implemented using two
+  ``get_partial_values`` operations on the underlying store, one for retrieving
+  the indices and one for retrieving existing chunks.
+
+* ``set_partial_values(key_start_values)``: For each referenced key, check if
+  all available chunks in a shard are referenced. In this case, a shard can be
+  constructed according to the `Binary shard format`_ directly. For all other
+  keys, request the indices from the underlying store using
+  ``get_partial_values``. All chunks that are not updated completely and exist
+  according to the index (index offset and nbytes are both ``2^64 - 1``) need to
+  be read via ``get_partial_values`` from the underlying store. For
+  simplification purposes a shard may also be read completely, combining the
+  previous two `get` operations into one. Based on the existing chunks and value
+  ranges that need to be updated new shards are constructed according to the
+  `Binary shard format`_. All shards that need to be updated must now be set via
+  ``set`` or ``set_partial_values(key_start_values)``, depending on the chosen
+  writing strategy provided by the implementation. Specialized store
+  implementations that allow appending to a storage object may only need to read
+  the index to update it.
+
+* ``erase_values(keys)``: For each referenced key, check if all available chunks
+  in a shard are referenced. In this case, the full shard is removed using
+  ``erase_values`` on the underlying store. For all other keys, request the
+  indices from the underlying store using ``get_partial_values``. Update the
+  index using an offset and nbytes of ``2^64 - 1`` to mark missing chunks. The
+  updated index may be written in-place using
+  ``set_partial_values(key_start_values)``, or a larger rewrite of the shard may
+  be done including the index update, but also removing value ranges
+  corresponding to the erased chunks.
+
+* ``erase_prefix()``: If the prefix contains a part of the chunk-grid key, this
+  part is translated to the referenced shard and contained chunks. For affected
+  shards where all contained chunks are erased the prefix is rewritten to the
+  corresponding shard key and the operation is relayed to the underlying store.
+  For all shards where only some chunks are erased the affected chunks are
+  removed by invoking the operation ``erase_values`` on this storage transformer
+  with the respective chunk keys.
+
+* ``list()``: See ``list_prefix`` with the prefix ``/``.
+
+* ``list_prefix(prefix)``: If the prefix contains a part of the chunk-grid key,
+  this part is translated to the referenced shard and contained chunks. Then,
+  ``list_prefix`` is called on the underlying store with the translated prefix.
+  For all listed shards request the indices from the underlying store using
+  ``get_partial_values``. Existing chunks, where the index offset or nbytes are
+  not ``2^64 - 1`` are then listed by their original key.
```

Comment:
Just noting that, if a read of each chunk needs to be done to list a prefix, this simple list operation could get pretty expensive for large arrays. If the arrays rarely change, this is another reason to have some level of additional metadata caching, as explored in #154.

Reply:
I also was considering that there might be a "less strict" implementation for the list operations, that might also list unavailable keys. Do you think that's a viable option? I guess that might just be a detail a specific implementation might offer, but it could still be mentioned here.

Reply:
Ultimately these operations are basically just examples of something an implementation might make use of, rather than an actual part of the spec that implementations need to be aware of.  So I don't think we really need to be too concerned about the efficiency of these particular operations, as long the actual sharded zarr array can be used efficiently.

In general I see two uses for list operations:
- When deleting a region of an array, we need to know which shards can be deleted entirely and which might need to be partially modified.  This would benefit from knowledge of the sharding hierarchy, as this isn't quite the same as either a "strict" or "non strict" list operation.
- When trying to efficiently operate on an array that has been written sparsely, i.e. only some chunks have been written, you might want to determine which shards are present, and for those shards, which chunks are present.  This could potentially just make use of a generic but strict "list" operation.

Do you have a use in mind for a "less strict" list operation that would just include all chunks for any shard that is present?

Most likely in Neuroglancer and TensorStore I won't actually have "storage transformers" as a concept; instead, I'll just treat sharding specially.